keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1474 @@
+  serviceAccountName [2] UTF8String,
+  serviceAccountUID [3] UTF8String,
+  podName [4] UTF8String,
+  podUID [5] UTF8String,
+  nodeName [6] UTF8String,
+  nodeUID [7] UTF8String
+}
+```
+
+Utility functions for embedding and extracting this extension from `crypto/x509`
+Certificate objects are defined in `k8s.io/component-helpers/kubernetesx509`.
+
+This extension, and the utility functions for manipulating it, are developed for
+use by core Kubernetes components, although third parties could make use of
+them.  The extension should only be changed in backwards-compatible ways.
+
+#### API Server PodIdentity Extension Support
+
+The kube-apiserver is extended to authenticate TLS client certificates that
+represent service account identities in the following format:
+
+* Are signed by root certificates provided by a new `--pod-client-ca-file` flag
+  (with automatic reload when the contents of the file change),
+* Contain no SubjectAlternateName data, and
+* Contain an X.509 PodIdentity extension.
+
+These certificates are translated to an authenticator.Response that has:
+* `User.Name` set to
+  `system:serviceaccount:{PodIdentity.Namespace}:{PodIdentity.ServiceAccountName}`.
+* `User.UID` set to `PodIdentity.ServiceAccountUID`.
+* `User.Extra["authentication.kubernetes.io/pod-name"]` set to
+  `{PodIdentity.PodName}`
+* `User.Extra["authentication.kubernetes.io/pod-uid"]` set to
+  `{PodIdentity.PodUID}`
+* `User.Extra["authentication.kubernetes.io/node-name"]` set to
+  `{PodIdentity.NodeName}`
+* `User.Extra["authentication.kubernetes.io/node-uid"]` set to
+  `{PodIdentity.NodeUID}`
+* `User.Extra["authentication.kubernetes.io/credential-id"]` set to
+  `X509SHA256=<sha256-hash-of-the-certificate>`.
+
+The net effect is that authenticating via a PodIdentity-bearing cert has the
+same object-bound behavior as authenticating via a bound service account token.
+
+During Alpha, this feature will be gated by the
+`PodCertificateKubeApiserverClientAcceptance` feature gate.
+
+#### API Server Pod Client Certificate Signer
+
+The kube-controller-manager is extended to support a new certificate signer,
+`kubernetes.io/kube-apiserver-client-pod`.
+
+The controller for this signer name automatically issues certificates for
+PodCertificateRequests that are addressed to it.  It performs no additional
+validation on PodCertificateRequests beyond the validation enforced by
+kube-apiserver.
+
+The certificates it issues are meant to be understood by kube-apiserver as
+described above.  They:
+* Have a PodIdentity extension (critical) that contains the information from
+  `PodCertificateRequest.Spec`.
+* Key usages consistent with the certificate being used as a TLS client
+  certificate.
+* It uses a CA specified by a new pair of private key / root certificate flags
+  (although implementations are free to set this flag to point at the CA as
+  another signer).
+* It issues certificates that are valid for 24 hours, and tells Kubelet to begin
```

Comment:
What if a client asks for a shorter TTL?  What is the min TTL this signer supports?

Reply:
Good point.  I will update so that it issues certs for `min(maxExpirationSeconds, 24hours)`.